File: Typescript/Tree/TreeADT.ts
import { Comparable } from "./BinarySearchTree";
import TraverseSearchTree from "./TraverseSearchTree";
import NODE from "./Node";

export enum TreeTraverseType {
  PreOrder,
  InOrder,
  PostOrder,
  LevelOrder,
}

export interface TreeADT<T extends Comparable<T>> {
  isEmpty(): boolean;
  size(): number;
  height(): number;
  contains(element: T): boolean;
  add(element: T): boolean;
  remove(element: T): boolean;
  traverse(type: TreeTraverseType): TraverseSearchTree<T>;
}

File: Java/src/Tree/TreeADT.java
package Tree;

import java.util.Iterator;

public interface TreeADT<T> {
    boolean isEmpty();
    int size();
    int height();
    boolean contains(T elem);
    boolean add(T elem);
    boolean remove(T elem);
    Iterator<T> traverse(TreeTraverseType type);
}
